Use `git diff-tree` for `DiffFileTree` on diff pages

McRaeAlex opened the pull request:
Modify Diff View FileTree to show all files

## Changes

* removes Show Status button on diff
* uses `git diff-tree` to generate the file tree for the diff
* doesn't reload the diff tree each time we load more files in the preview
* selecting and unloaded file will keep loading until that file is loaded
* removes `DiffFileList.vue` and "Show Stats" in diff options

## Open Questions

* selecting and unloaded file will keep loading until that file is loaded. Is this behaviour okay? It matches what github does.

### Demo

In this demo I set `git.MAX_GIT_DIFF_FILES=1` in my `app.ini` to demonstrate a worst case example. In most cases the behaviour isn't nearly as jarring as we load a bunch of files at a time.

https://github.com/user-attachments/assets/72f29663-d6fc-472d-94fa-7fb5950c2836

McRaeAlex commented:
@delvh @lunny I taken this out of draft so we can discuss the open question and any other behaviour you want to go over.

kerwin612 commented on templates/repo/diff/options_dropdown.tmpl:
```diff
@@ -1,7 +1,6 @@
 <div class="ui dropdown tiny basic button" data-tooltip-content="{{ctx.Locale.Tr "repo.diff.options_button"}}">
 	{{svg "octicon-kebab-horizontal"}}
 	<div class="menu">
-		<a class="item" id="show-file-list-btn">{{ctx.Locale.Tr "repo.diff.show_diff_stats"}}</a>
```
It is necessary to check whether "repo.diff.show_diff_stats" in the `locale` needs to be deleted synchronously.

McRaeAlex replied:
Good point it looks like this was the only place `show_diff_stats` is used. Will remove.

lunny commented:
I think you don't need to update all the language files except en_US.

McRaeAlex commented:
> I think you don't need to update all the language files except en_US.

I am not sure I understand, If I remove it from `en_US` then why wouldn't I also remove it from every other language?

lunny commented:
> > I think you don't need to update all the language files except en_US.
> 
> I am not sure I understand, If I remove it from `en_US` then why wouldn't I also remove it from every other language?

Because other language files will be synced from translate.gitea.com, so basicaly you just need to change the `en_US` file.

lunny commented:
I did a test manually, looks like a modified file can not be selected.

McRaeAlex commented:
@lunny Apologizes for the delay.

I wasn't able to reproduce your issue. Could you provide some more details?

https://github.com/user-attachments/assets/ec8f4836-f6de-4b74-b417-046bc562e967

delvh commented on routers/web/repo/pull.go:
```diff
@@ -816,6 +818,27 @@ func viewPullFiles(ctx *context.Context, specifiedStartCommit, specifiedEndCommi
+		if shouldGetUserSpecificDiff {
+			// This sort of sucks because we already fetch this when getting the diff
+			review, err := pull_model.GetNewestReviewState(ctx, ctx.Doer.ID, issue.ID)
+			if !(err != nil || review == nil || review.UpdatedFiles == nil) {
```
```suggestion
			if err == nil && review != nil && review.UpdatedFiles != nil {
```

delvh commented on web_src/js/components/DiffFileTreeItem.vue:
```diff
@@ -48,27 +35,37 @@ function fileIcon(file: File) {
-    <span class="gt-ellipsis">{{ item.name }}</span>
-  </div>
 
-  <div v-if="item.children?.length" v-show="!collapsed" class="sub-items">
```
What happened to the `v-if`?

delvh commented on web_src/js/features/repo-diff.ts:
```diff
@@ -234,17 +232,83 @@ function initRepoDiffShowMore() {
+    return true; // The element is already loaded
+  }
+
+  if (!pageData.diffFileInfo.isIncomplete) {
+    return true; // All files are already loaded
+  }
+
+  return false;
```
```suggestion
  return !pageData.diffFileInfo.isIncomplete;
```

delvh commented on web_src/js/features/repo-diff.ts:
```diff
@@ -234,17 +232,83 @@ function initRepoDiffShowMore() {
+    return true; // No hash to look for
+  }
+
+  const targetElement = document.querySelector(hash);
```
Hmm…
That is one of the rare instances where it is probably better to use `document.getElementById` instead of `querySelector`.

Originally, I was uncertain if I should really post this comment, then I scrolled down to see a `hash change listener`.

In other words:
If the user manually modifies the hash, it automatically loads incomplete diffs until the diff has been fully loaded.
Even worse, since the user can change the hash themselves, nothing prevents them from entering a `.` or similar, which can produce false positives with the query selector.

So, all in all I'm wondering if that is actually a nice shortcut for power users: Want to load all files in the PR? Simply add `#a` to the PR, and boom, all files are loaded.

On the other hand however, we have the reason why the file limits were introduced in the first place: Many people run Gitea on cheap hardware.
If we allow querying files infinitely, it can overwhelm cheap servers.

But on the other hand, the performance gain is probably worth it.

delvh replied:
So overall, we can probably keep it as it is but should only use `getElementById` instead to prevent false positives

McRaeAlex commented on web_src/js/components/DiffFileTree.vue:
```diff
@@ -78,8 +21,8 @@ onMounted(() => {
 
-  hashChangeListener();
   window.addEventListener('hashchange', hashChangeListener);
+  nextTick(hashChangeListener);
```
Mistake. I have reverted the code.

wxiaoguang commented on web_src/js/features/repo-diff.ts:
```diff
@@ -184,8 +184,6 @@ export async function loadMoreFiles(url: string) {
     // 1. append the newly loaded file list items to the existing list
     $('#diff-incomplete').replaceWith($resp.find('#diff-file-boxes').children());
-    // 2. re-execute the script to append the newly loaded items to the JS variables to refresh the DiffFileTree
-    $('body').append($resp.find('script#diff-data-script'));
```
Since the legacy `script#diff-data-script` won't be executed again anymore, these JS code should be moved from "tmpl" to "ts" files?

McRaeAlex replied:
I am not sure what you mean, the script embedded in the template still runs on first load and needs information from go to template the script.

It could be more minimal such that the template only embeds JSON information however this was also true before this pull request.

Let me know if I have misunderstood. 😃

wxiaoguang replied:
> It could be more minimal such that the template only embeds JSON information however this was also true before this pull request.

Actually that was not true before this PR .... the old code was written intentionally to make it could reload the data when "load more", but it seems that the "load more" won't affect the script in tmpl now.

Never mind, I could do some following up changes to refactor the legacy code.

wxiaoguang replied:
Done in 7adbb7670f1be7b1a892a744e1ce1b271812740c

wxiaoguang commented on web_src/js/features/repo-diff.ts:
```diff
@@ -234,17 +232,80 @@ function initRepoDiffShowMore() {
+  }
+
+  // eslint-disable-next-line unicorn/prefer-query-selector
+  const targetElement = document.getElementById(hash.substring(1));
```
Still prefer `querySelector` but not `getElementById`

eg: `document.querySelector(hash)` ?

delvh replied:
I know, but in this case, the result can be controlled by the user (look below, there's a hashchange listener).
If we don't use getElementById explicitly, false positives/negatives could happen.

wxiaoguang replied:
Why? What's the case to cause "false positives/negatives could happen"?

delvh replied:
The user manually changes the page hash to, I don't know `#file-input%20.ui.button` or something like that.

delvh replied:
Which, if my guess is correct, can lead to the false element being selected

wxiaoguang replied:
But you can also input something like `#navbar` `#mobile-notifications-icon` `#navbar-expand-toggle` `#issue-title-display` and a lot of others to make `getElementById` select other "false elements" too, right?

What's the difference?

delvh replied:
True.
The difference would be that at least this weird case would then be prevented.
However, I do agree that users doing something like that probably deserve whatever they get.

delvh replied:
In that case, `querySelector` should be fine.

wxiaoguang replied:
Hmm, I think I found an edge case which makes `getElementById` win ..... `document.querySelector('#')` throws JS exception 🤣  If we don't want to use `try-catch` .... then let's keep the original `getElementById` and add some comments.

Thank you thoughtful suggestion  👍

silverwind commented on web_src/js/features/repo-diff.ts:
```diff
@@ -234,17 +232,80 @@ function initRepoDiffShowMore() {
 }
 
+function shouldStopLoading() {
+  const hash = window.location.hash;
```
```suggestion
  const {hash} = window.location;
```

Same, but shorter.

silverwind commented on web_src/js/features/repo-diff.ts:
```diff
@@ -234,17 +232,80 @@ function initRepoDiffShowMore() {
 }
 
+function shouldStopLoading() {
+  const hash = window.location.hash;
+
+  if (!hash) {
+    return true; // No hash to look for
+  }
+
+  // eslint-disable-next-line unicorn/prefer-query-selector
+  const targetElement = document.getElementById(hash.substring(1));
```
```suggestion
  if (!window.location.hash) {
    return true; // No hash to look for
  }

  // eslint-disable-next-line unicorn/prefer-query-selector
  const targetElement = document.getElementById(window.location.hash.substring(1));
```

No need for that variable imho.

silverwind commented on web_src/js/components/file_tree.ts:
```diff
@@ -0,0 +1,90 @@
+
+  let parentItem = pathToItem.get(dirParentPath);
+  if (!parentItem) {
+        // if the parent node does not exist, create it
```
```suggestion
    // if the parent node does not exist, create it
```

silverwind commented on web_src/js/components/file_tree.ts:
```diff
@@ -0,0 +1,90 @@
+function splitPathLast(path: string): [string, string] {
+  const lastSlash = path.lastIndexOf('/');
+  return [path.substring(0, lastSlash), path.substring(lastSlash + 1)];
+}
```
This can be replaced by `basename` and `dirname` functions. The former already exists, the latter is easy to add:

https://github.com/go-gitea/gitea/blob/38ccc8e3e48e9b15f175287c0721ce0ff9892e5a/web_src/js/utils.ts#L5

silverwind commented on web_src/js/components/file_tree.ts:
```diff
@@ -0,0 +1,90 @@
+      dir.children = child.children;
+    }
+  }
+}
```
This file is not exporting any components, it shouldn't be in `web_src/js/components`. I'd put it in `web_src/js/utils/filetree.ts` instead.

silverwind commented on web_src/js/components/file_tree.ts:
```diff
@@ -0,0 +1,90 @@
+    path: string;
+
+    // we disable eslint here because we have a recursive type definition
+    // eslint-disable-next-line no-use-before-define
```
This is likely a configuration issue, we should be using `@typescript-eslint/no-use-before-define` instead which should not raise this issue. I will check later.

https://typescript-eslint.io/rules/no-use-before-define/

silverwind replied:
Will be fixed with https://github.com/go-gitea/gitea/pull/33715 after which this rule disablement can be removed.

silverwind commented on web_src/js/components/file_tree.ts:
```diff
@@ -0,0 +1,90 @@
+    isFile: false;
+    name: string;
+    path: string;
+
+    // we disable eslint here because we have a recursive type definition
+    // eslint-disable-next-line no-use-before-define
```
```suggestion
```

wxiaoguang commented:
Still have question about the script block. Are these variables used? Is the comment correct?

![image](https://github.com/user-attachments/assets/f251820a-1b9a-4601-a68a-936603ee0383)

wxiaoguang commented:
ps: no need to force-push when review starts, the final merge is a squash merge. Force-push makes the review difficult because every-line needs to be re-reviewed since everything is in a commit.

wxiaoguang commented on routers/web/repo/treelist.go:
```diff
@@ -52,3 +54,40 @@ func isExcludedEntry(entry *git.TreeEntry) bool {
+	for _, file := range diffTree.Files {
+		nameHash := git.HashFilePathForWebUI(file.HeadPath)
+		isSubmodule := file.HeadMode == git.EntryModeCommit
+		isBinary := file.HeadMode == git.EntryModeExec
```
It's not true, "exec" doesn't mean "binary". And I do not see where "IsBinary" is really used.

McRaeAlex replied:
Removed

wxiaoguang commented on routers/web/repo/treelist.go:
```diff
@@ -52,3 +54,40 @@ func isExcludedEntry(entry *git.TreeEntry) bool {
+	files := make([]FileDiffFile, 0, len(diffTree.Files))
+
+	for _, file := range diffTree.Files {
+		nameHash := git.HashFilePathForWebUI(file.HeadPath)
```
Why "HeadPath"? For a deleted file, does "HeadPath" exist?

McRaeAlex replied:
HeadPath exists for added, removed and changed files. It differs from basePath when a rename is detected.

wxiaoguang replied:
What's the difference from the `filesViewedState[file.Path()]` below? Why here only uses "head path" but `filesViewedState` uses "base path" or "head path"?

wxiaoguang commented on routers/web/repo/treelist.go:
```diff
@@ -52,3 +54,40 @@ func isExcludedEntry(entry *git.TreeEntry) bool {
+		isSubmodule := file.HeadMode == git.EntryModeCommit
+		isBinary := file.HeadMode == git.EntryModeExec
+
+		isViewed := false
+		if fileViewedState, ok := filesViewedState[file.Path()]; ok {
+			isViewed = (fileViewedState == pull_model.Viewed)
+		}
```
According to Golang spec, accessing nil map is safe:

```
isViewed := filesViewedState[file.Path()] == pull_model.Viewed
```

McRaeAlex replied:
TIL 👍

McRaeAlex commented:
> ps: no need to force-push when review starts, the final merge is a squash merge. Force-push makes the review difficult because every-line needs to be re-reviewed since everything is in a commit.

I use [jj](https://github.com/jj-vcs/jj) which makes this behaviour automatic and keeps my "stacked" changes up to date. I can switch to using the git cli and adding commits if its preferred.

> Still have question about the script block. Are these variables used? Is the comment correct?

I have updated the comment. `window.config.pageData.diffFileInfo` is used in `web_src/js/modules/stores.ts`. I cannot find any place where `diffData` is used (other than `diffData.isIncomplete`) for example `diffData.statisticsMessage` seems to be dead code, I have removed it.

wxiaoguang commented on routers/web/repo/treelist.go:
```diff
@@ -52,3 +54,33 @@ func isExcludedEntry(entry *git.TreeEntry) bool {
+	for _, file := range diffTree.Files {
+		nameHash := git.HashFilePathForWebUI(file.HeadPath)
+		isSubmodule := file.HeadMode == git.EntryModeCommit
+		isViewed := filesViewedState[file.Path()] == pull_model.Viewed
```
Still the question here: why use `file.Path()` to check `isViewed`?

If I understand correctly (by https://github.com/go-gitea/gitea/pull/33514/files#r1970735341): HeadPath always exists? So no need to introduce `func Path()`?

McRaeAlex replied:
Removed

silverwind commented on web_src/js/components/DiffFileTreeItem.vue:
```diff
@@ -2,21 +2,7 @@
-  file?: File;
-  children?: Item[];
-};
+import {type Item, type File, type FileStatus} from '../utils/filetree.ts';
```
```suggestion
import type {Item, File, FileStatus} from '../utils/filetree.ts';
```

silverwind commented on templates/repo/diff/box.tmpl:
```diff
@@ -57,28 +57,6 @@
-			diffFileInfo.files.push(...diffDataFiles);
-			window.config.pageData.diffFileInfo = diffFileInfo;
-		</script>
-		<div id="diff-file-list"></div>
```
With this removal, will it not flash wrong content on page load when diff tree is visible?

silverwind replied:
Ah I may be misunderstanding and this was just some data modification that has no reason to be in templates.

wxiaoguang replied:
Yes, it's not related to "content flash" (and this block was marked as "module", it only executes after DOM ready).

Some old PRs used very tricky methods to make the Vue work, now it's the time to make code overall right.

silverwind replied:
Yeah we should remove all such template js, except those cases that serve the purpose of avoiding flash, but those are better migrated to web components.

wxiaoguang commented:
Found a bug:

Now the file tree doesn't follow the selected commit:

![image](https://github.com/user-attachments/assets/539c5675-e5af-487c-8537-2e55cdedcd31)
